content/6_Advanced/Critical.mdx

```diff
@@ -1,26 +1,526 @@
+* Let $e(u)$ be the entry time in node $u$ doing a [Euler tour](/gold/tree-euler).
+* Let the **semi-dominator** of $u$, or $\texttt{sdom}(u)$, be a $v$ such that there's a path from $v$ to $u$
+  and $e(i) \ge e(u)$ for every intermediate node $i$ along the path from $u$ to $v$, excluding the ends.
+  If there's multiple nodes that satisfy this requirement, we take the node $v$ with the smallest $e(v)$.
+* Let the **relative dominator** of $u$, or $\texttt{rdom}(u)$, be the vertex $x \ne \texttt{sdom}(u)$ on the path from $\texttt{sdom}(u)$ to $u$ in the Euler tour tree with the lowest sdom node number.
+  Unlike with the sdom, ties in this function can be broken arbitrarily.
+
+A dominator tree is a tree where each node's children are those nodes it immediately dominates. The start node is the root of the tree.
+
+Given this definition, we can see that if a node dominates another, then the former is
+an ancestor of the latter in the dominator tree.
+Thus, the answer to the CSES problem is the set of all nodes that lie on the path from the root to node $n$ in the tree.
+
+The following graph shows the sdom for every node.
+The full-color edges represent the edges part of the DFS tree.
+
+![Graph2](./assets/dom_tree_sdom.png)
+
+### Important properties
+
+Proofs of these properties are located later in the module.
+For all of these properties, let $u$ be a node that isn't the starting node.
+
+1. $\texttt{sdom}(u)$ is a proper ancestor of $u$ is the DFS tree.
+2. If $\texttt{rdom}(u)=u$, then $\texttt{idom}(u)=\texttt{sdom}(u)$.
+3. If not, then $\texttt{idom}(u)=\texttt{idom}(\texttt{rdom}(u))$
+
+## Algorithm Overview
+
+Before we get into the nitty-gritty, here's a brief outline of how exactly we're going to build up this dominator tree.
+
+1. Compute the sdom of every node besides the start.
+2. Compute the *rdom* of every node besides the start.
+3. Visit all the vertices in the DFS tree and calculate their immediate dominator using the second and third properties that were listed above.
+   Notice that due to how we defined the rdom of a node, a preorder traversal will always visit the rdom of a node before the node itself if the two aren't the same.
+
+The first and second steps are awfully vague- let's clear those up now, shall we?
+
+### Computing $\texttt{sdom}$
+
+We can compute $\texttt{sdom}(u)$ as the minimum node in the intersection of the following groups:
+1. All the nodes $y$ such that there's an edge from $y$ to $u$ and $e(y) \le e(u)$.
+2. All the values of $\texttt{sdom}(x)$ where $x$ is any node such that there's an edge from $u$ to $x$ and $e(x) > e(u)$.
+   To be more mathematically precise, we can define this group as
+   $$
+   \{\texttt{sdom}(x)\ |\ (u, x) \in E\text{ and }e(x)>e(u)\}
+   $$
+
+The proof of why this works is beyond the scope of this module.
```

Comment:
is there a link where ppl can learn more?

Reply:
Ok I added a link